tools/logparser2.py

```diff
@@ -11,86 +11,144 @@
 import sys
 from zipfile import ZipFile
 
+def _meta(log):
+    with log.open(log.metaname) as stream:
+        for line in filter(lambda a: ':' in a, stream):  # skip unknown lines:
+            sensor, filepath = line.strip().split(':')
+            yield sensor, os.path.basename(filepath)
 
-def timestamps_gen(id, stream):
+def _stream(log, id):
     if id == 'can': # binary log, not implemented
         return
+
     if id == 'timestamps':
-        for line in stream:
-            line = ast.literal_eval(line)
-            assert len(line) == 3
-            yield float(line[0]), 'can', line[1:]
-    for i, line in enumerate(stream):
-        if i % 2 == 1:
-            data = ast.literal_eval(line)
-            yield timestamp, id, data
-        else:
-            vals = line.split()
-            if len(vals) != 2: # only at the end of file
-                return
-            timestamp = float(vals[1])
-
-
-def sensor_gen(meta_filename, selected=None):
+        with log.open(log.streams[id]) as stream:
+            for line in stream:
+                line = ast.literal_eval(line)
+                assert len(line) == 3
+                yield float(line[0]), 'can', line[1:]
+        return
+
+    with log.open(log.streams[id]) as stream:
+        for i, line in enumerate(stream):
+            if i % 2 == 1:
+                data = ast.literal_eval(line)
+                yield timestamp, id, data
+            else:
+                vals = line.split()
+                if len(vals) != 2: # only at the end of file
+                    return
+                timestamp = float(vals[1])
+
+
+class DirLog:
+    def __init__(self, metapath):
+        self.dirpath = os.path.dirname(metapath)
+        self.metaname = os.path.basename(metapath)
+
+    def __enter__(self):
+        self.streams = dict(_meta(self))
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def open(self, filename):
+        assert '/' not in filename
+        return open(self.dirpath + '/' + filename)
+
+    def read(self, filename):
+        with self.open(filename) as f:
+            return f.read()
+
+class ZipLog:
+    def __init__(self, filepath):
+        self.filepath = filepath
+
+    def __enter__(self):
+        self.zip = ZipFile(self.filepath)
+        meta_filename = [name for name in self.zip.namelist() if name.startswith('meta_')]
+        assert len(meta_filename) == 1
+        self.metaname = meta_filename[0]
+        self.streams = dict(_meta(self))
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.zip.close()
+
+    def open(self, filename):
+        assert '/' not in filename
+        return self.zip.open(filename)
+
+    def read(self, filename):
+        assert '/' not in filename
+        return self.zip.read(filename)
+
+
+def Log(path):
```

Comment:
hm, do we want to have functions with capital letters? Originally I expected that this is class ... I know it is kind of factory ... hmm, not sure

Reply:
Yeah, I was not sure of this one myself. Let me know if you become sure :wink:

Reply:
OK, keep `Log` for now, but please add comment that it actually factory function returning different classes